File: 5_software_and_its_engineering/1_software_notations_and_tools/common_programming_problems/0_declarative_model/arrays/1.1.1.StringUniqueCharacters-0.scala

def contain_all_unique_characters(string: String): Boolean = {
    string.distinct.length() == string.length()
}

@main def main(): Unit = {
    val s1 = "gofreetech"  // false
    println(contain_all_unique_characters(s1))

    val s2 = "python"  // true
    println(contain_all_unique_characters(s2))

    val s3 = "p"  // true
    println(contain_all_unique_characters(s3))

    val s4 = ""  // true
    println(contain_all_unique_characters(s4))
}

File: 5_software_and_its_engineering/0_software_notations_and_tools/unified_discipline_of_computer_programming/programs_per_computational_model/3_stateful/arrays_and_strings/1.1.1.StringUniqueCharacters-0.py
def contain_all_unique_characters(s):
    sorted_version = sorted(s)
    string_length = len(s)
    last_index = string_length - 1
    for i in range(0, string_length):
        if i < last_index and sorted_version[i] == sorted_version[i + 1]:
            return False
    return True


s1 = "gofreetech"  # false
print(contain_all_unique_characters(s1))

s2 = "python"  # true
print(contain_all_unique_characters(s2))

s3 = "p"  # true
print(contain_all_unique_characters(s3))

s4 = ""  # true
print(contain_all_unique_characters(s4))


"""
    Performance
        N = length of string

        Time = O( N log N + N)
            sort takes = N log N
        Space = 𝑂(𝑁)
            The space cost of the sorted array
"""
